telemetry/firmware/telemetry.c

```diff
@@ -0,0 +1,353 @@
+/*
+   Header:
+    This is the functional code for the telemetry board
+    Function: Responsible for pulling safety & failure critical information from CANBus and passing it along to UART for telemetry RF module
+            : Interface with RF Module 
+            : Interface with CANBus
+
+
+    confirm UART setup
+    confirm CAN setup
+    setup UART sending  
+
+   Author:
+    adi ramachandran
+    aramachandran@olin.edu
+    6307969467
+        
+ */
+
+
+/*----- Includes -----*/
+#include <stdio.h>
+#include <stdlib.h>
+#include "can_api.h"
+#include "queue.h"
+
+/*----- Macro Definitions -----*/
+
+
+/* Defining queue structures
+- lengths are sort of arbitrary 
+*/ 
+#define UART_QUEUE_SIZE                 512 // 512 bytes too long? 
+
+
+// gFlag positions
+#define UART_SEND_READY                 0
+#define UART_RECIEVED                   1
+
+
+// CAN Mailboxes
+#define THROTTLE_CMD_MBOX                  0
+#define THROTTLE_BRD_MBOX                  1
+#define BRAKE_MBOX                         2
+#define AIR_CONTROL_SENSE_MBOX             3
+#define BMS_CORE_MBOX                      4
+#define CELL_VOLT_TEMP_MBOX                5
+
+#define CAN_MSK_6                           0b11000000 // custom mask for cell temps & voltages
+// handle another 15 mailboxes lol 
+
+
+// UART 
+
+// values within LINCR register
+#define ENABLE_LIN_AND_UART             LENA // can i leave this as LENA or does it have to be bit 3 (page 190)
+#define ENABLE_UART                     LCMD2
+#define ENABLE_RX                       LMCD1
+#define ENABLE_TX                       LCMD0
+
+#define UART_DATA                       LINDAT             // lin data register
+#define UART_BAUD                       9600
+
+
+// shutdown_error_flag codes within error var
+
+#define BSPD_FAIL                       0
+#define INERTIA_SWITCH_FAIL             1
+#define BOTS_FAIL                       2
+#define COCKPIT_ESTOP_FAIL              3
+#define HV_DISCONNECT_FAIL              4
+#define HVD_CONN_FAIL                   5
+#define PACK_CONN_FAIL                  6
+#define BMS_FAIL                        7
+#define IMD_FAIL                        8
+#define TSMS_FAIL                       9
+
+
+/*----- Global Variables -----*/
+
+volatile uint8_t global_flag = 0x00; // global flag
+
+volatile uint16_t shutdown_error_flag = 0x00; // system failure error code flag - default 0x00. 
+
+volatile uint8_t can_recv_msg[8] = {0}; // CAN msg recieved
+
+
+QUEUE(UART, uint8_t, UART_QUEUE_SIZE); 
+voltatile struct queue_UART UART_queue; // creating a UART_queue of type queue_UART 
+
+
+
+
+
+/*----- Interrupt(s) -----*/
+
+
+ISR(CAN_INT_vect){ 
+
+    CANPAGE = (THROTTLE_BRD_MBOX<<MOBNB0); 
```

Comment:
Datasheet 20.11.6/7, consider using the `CANIE` and `CANSIT` registers to tell which MOb triggered the interrupt.